File: core/core-web/src/main/java/ch/difty/scipamato/core/web/common/BasePanel.java
package ch.difty.scipamato.core.web.common;

import org.apache.wicket.model.IModel;
import org.apache.wicket.model.StringResourceModel;
import org.apache.wicket.spring.injection.annot.SpringBean;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.security.core.context.SecurityContextHolder;

import ch.difty.scipamato.common.navigator.ItemNavigator;
import ch.difty.scipamato.common.web.AbstractPanel;
import ch.difty.scipamato.common.web.Mode;
import ch.difty.scipamato.common.web.ScipamatoWebSessionFacade;
import ch.difty.scipamato.core.config.ApplicationCoreProperties;
import ch.difty.scipamato.core.entity.User;

@SuppressWarnings("unused")
public abstract class BasePanel<T> extends AbstractPanel<T> {

    private static final long serialVersionUID = 1L;

    @SpringBean
    private ApplicationCoreProperties properties;
    @SpringBean
    private ScipamatoWebSessionFacade webSessionFacade;

    protected BasePanel(@NotNull final String id) {
        this(id, null, Mode.VIEW);
    }

    @SuppressWarnings("unused")
    BasePanel(@NotNull final String id, @Nullable IModel<T> model) {
        this(id, model, Mode.VIEW);
    }

    protected BasePanel(@NotNull final String id, @Nullable IModel<T> model, @NotNull Mode mode) {
        super(id, model, mode);
    }

    @NotNull
    protected ApplicationCoreProperties getProperties() {
        return properties;
    }

    @NotNull
    protected ItemNavigator<Long> getPaperIdManager() {
        return webSessionFacade.getPaperIdManager();
    }

    @NotNull
    protected String getLocalization() {
        return webSessionFacade.getLanguageCode();
    }

    /**
     * @return the currently active user
     */
    @NotNull
    protected User getActiveUser() {
        Object principal = SecurityContextHolder
            .getContext()
            .getAuthentication()
            .getPrincipal();
        return (User) principal;
    }

    /**
     * Retrieves the labels resource string.
     *
     * @param componentId
     *     the wicket id of the component the label is assigned to.
     * @return the label string, taken from the resource.
     */
    @NotNull
    protected String getLabelResourceFor(final String componentId) {
        return getResourceFor(componentId);
    }

    /**
     * Retrieves the labels resource string (short form).
     *
     * @param componentId
     *     the wicket id of the component the label is assigned to.
     * @return the label string, taken from the resource.
     */
    @NotNull
    protected String getShortLabelResourceFor(@NotNull final String componentId) {
        return getResourceFor(componentId);
    }

    private String getResourceFor(final String componentId) {
        return new StringResourceModel(componentId + AbstractPanel.LABEL_RESOURCE_TAG, this, null).getString();
    }
}

File: core/core-web/src/main/java/ch/difty/scipamato/core/web/common/BasePanel.kt
package ch.difty.scipamato.core.web.common

import ch.difty.scipamato.common.navigator.ItemNavigator
import ch.difty.scipamato.common.web.AbstractPanel
import ch.difty.scipamato.common.web.LABEL_RESOURCE_TAG
import ch.difty.scipamato.common.web.Mode
import ch.difty.scipamato.common.web.SHORT_LABEL_RESOURCE_TAG
import ch.difty.scipamato.common.web.ScipamatoWebSessionFacade
import ch.difty.scipamato.core.config.ApplicationCoreProperties
import ch.difty.scipamato.core.entity.User
import org.apache.wicket.model.IModel
import org.apache.wicket.model.StringResourceModel
import org.apache.wicket.spring.injection.annot.SpringBean
import org.springframework.security.core.context.SecurityContextHolder

abstract class BasePanel<T> @JvmOverloads constructor(
    id: String,
    model: IModel<T>? = null,
    mode: Mode = Mode.VIEW,
) : AbstractPanel<T>(id, model, mode) {

    @SpringBean
    protected lateinit var properties: ApplicationCoreProperties

    @SpringBean
    private lateinit var webSessionFacade: ScipamatoWebSessionFacade

    protected val paperIdManager: ItemNavigator<Long>
        get() = webSessionFacade.paperIdManager

    protected val activeUser: User
        get() = SecurityContextHolder.getContext().authentication.principal as User

    protected val localization: String
        get() = webSessionFacade.languageCode

    /**
     * Retrieves the label's resource string for the provided [componentId].
     */
    protected fun getLabelResourceFor(componentId: String): String = getResourceFor(componentId, LABEL_RESOURCE_TAG)

    /**
     * Retrieves the label's resource string (short form) for given [componentId].
     */
    protected fun getShortLabelResourceFor(componentId: String): String = getResourceFor(componentId, SHORT_LABEL_RESOURCE_TAG)

    private fun getResourceFor(componentId: String, tag: String): String =
        StringResourceModel("$componentId$tag", this, null).string

    companion object {
        private const val serialVersionUID = 1L
    }
}
